sklearn_numba_dpex/kmeans/kernels/utils.py

```diff
@@ -25,6 +25,150 @@
+
+    n_selected_gt_threshold = n_selected_gt_threshold - 1
+    zero = dtype(0.0)
+
+    @dpex.kernel
+    # fmt: off
+    def relocate_empty_clusters(
+        X_t,                        # IN READ-ONLY   (n_features, n_samples)
+        sample_weight,              # IN READ-ONLY   (n_samples,)
+        assignments_idx,            # IN             (n_samples,)
+        samples_far_from_center,    # IN             (n_samples,)
+        empty_clusters_list,        # IN             (n_clusters,)
+        per_sample_inertia,         # INOUT          (n_samples,)
+        centroids_t,                # INOUT          (n_features, n_clusters)
+        cluster_sizes               # INOUT          (n_clusters,)
+            ):
+    # fmt: on
+        """NB: because of how the array was created, samples_far_from_center values
+        are located between (n_selected_gt_threshold - relocated_idx) and
+        (relocated_idx) indices.
+        """
+        group_idx = dpex.get_group_id(0)
+        item_idx = dpex.get_local_id(0)
+        relocated_idx = group_idx // n_work_groups_for_cluster
+        feature_idx = ((group_idx % n_work_groups_for_cluster) * work_group_size) + item_idx
+
+        if feature_idx >= n_features:
+            return
+
+        relocated_cluster_idx = empty_clusters_list[relocated_idx]
+        new_location_X_idx = samples_far_from_center[n_selected_gt_threshold - relocated_idx]
+        new_location_previous_assignment = assignments_idx[new_location_X_idx]
+
+        new_centroid_value = X_t[feature_idx, new_location_X_idx]
+        new_location_weight = sample_weight[new_location_X_idx]
+        X_centroid_addend = new_centroid_value * new_location_weight
+
+        # Cancel the contribution to the updated centroids of the sample that was once
+        # assigned to new_location_previous_assignment but is now assigned to the
+        # cluster of the centroids that relocates to this sample
+        dpex.atomic.sub(
+            centroids_t,
+            (feature_idx, new_location_previous_assignment),
+            X_centroid_addend
+            )
+
+        # The relocated centroid has only one contribution now, which is the sample
+        # to which it has been relocated to
+        centroids_t[feature_idx, relocated_cluster_idx] = X_centroid_addend
+
+        # Likewise, we update the weights in the clusters
+        if feature_idx == zero:
+            per_sample_inertia[new_location_X_idx] = zero
+            dpex.atomic.sub(
+                cluster_sizes,
+                new_location_previous_assignment,
+                new_location_weight
+            )
+            cluster_sizes[relocated_cluster_idx] = new_location_weight
+
+    return relocate_empty_clusters[global_size, work_group_size]
+
+
+@lru_cache
+def make_select_samples_far_from_centroid_kernel(
+    threshold, n_selected, n_samples, work_group_size
+):
+    global_size = math.ceil(n_selected / work_group_size) * work_group_size
+    zero = np.int64(0)
```

Comment:
From all the scalars being defined here, why is `zero` the only to be a `int64`? Is is for indexing reasons?

Reply:
I like expliciting the types because I'm afraid that else it could give unpractical types that would be implicitly casted during execution.

Maybe i've found that in `dpex.atomic` calls the index must be given in `int64` ? (else it fails ?) not sure anymore

:shrug: or maybe I should just drop expliciting the dtype of the constant and use plain numbers in the kernels.

Reply:
I was curious whether `np.int32` is usable instead.

> Maybe i've found that in dpex.atomic calls the index must be given in int64 ? (else it fails ?) not sure anymore

I suspect this this is the main reason.

> or maybe I should just drop expliciting the dtype of the constant and use plain numbers in the kernels.

I also like explicit types. +0 on changing to plain numbers.